cmd/agent/app/httpserver/server.go

```diff
@@ -0,0 +1,184 @@
+		handler.serveSamplingHTTP(w, r, false /* thriftEnums092 */)
+	})
+	mux.HandleFunc("/baggage", func(w http.ResponseWriter, r *http.Request) {
+		handler.serveBaggageHTTP(w, r)
+	})
+	return &http.Server{Addr: hostPort, Handler: mux}
+}
+
+func newHTTPHandler(manager Manager, mFactory metrics.Factory) *httpHandler {
+	handler := &httpHandler{manager: manager}
+	metrics.Init(&handler.metrics, mFactory, nil)
+	return handler
+}
+
+type httpHandler struct {
+	manager Manager
+	metrics struct {
+		// Number of good sampling requests
+		SamplingRequestSuccess metrics.Counter `metric:"http-server.requests" tags:"result=ok,type=sampling"`
+
+		// Number of good sampling requests against the old endpoint / using Thrift 0.9.2 enum codes
+		LegacySamplingRequestSuccess metrics.Counter `metric:"http-server.requests" tags:"result=ok,type=sampling-legacy"`
+
+		// Number of good baggage requests
+		BaggageRequestSuccess metrics.Counter `metric:"http-server.requests" tags:"result=ok,type=baggage"`
+
+		// Number of bad requests (400s)
+		BadRequest metrics.Counter `metric:"http-server.requests" tags:"result=err,status=4xx"`
+
+		// Number of collector proxy failures
+		TCollectorProxyFailures metrics.Counter `metric:"http-server.requests" tags:"result=err,status=5xx,type=tcollector-proxy"`
+
+		// Number of bad responses due to malformed thrift
+		BadThriftFailures metrics.Counter `metric:"http-server.requests" tags:"result=err,status=5xx,type=thrift"`
+
+		// Number of failed response writes from http server
+		WriteFailures metrics.Counter `metric:"http-server.requests" tags:"result=err,status=5xx,type=write"`
+	}
+}
+
+func (h *httpHandler) serviceFromRequest(w http.ResponseWriter, r *http.Request) (string, error) {
+	services := r.URL.Query()["service"]
+	if len(services) != 1 {
+		h.metrics.BadRequest.Inc(1)
+		http.Error(w, "'service' parameter must be provided once", http.StatusBadRequest)
+		return "", errBadRequest
+	}
+	return services[0], nil
+}
+
+func (h *httpHandler) writeJSON(w http.ResponseWriter, json []byte) error {
+	w.Header().Add("Content-Type", mimeTypeApplicationJSON)
+	if _, err := w.Write(json); err != nil {
+		h.metrics.WriteFailures.Inc(1)
+		return err
+	}
+	return nil
+}
+
+func (h *httpHandler) serveSamplingHTTP(w http.ResponseWriter, r *http.Request, thriftEnums092 bool) {
+	service, err := h.serviceFromRequest(w, r)
+	if err != nil {
+		return
+	}
+	resp, err := h.manager.GetSamplingStrategy(service)
+	if err != nil {
+		h.metrics.TCollectorProxyFailures.Inc(1)
+		http.Error(w, fmt.Sprintf("tcollector error: %+v", err), http.StatusInternalServerError)
+		return
+	}
+	jsonBytes, err := json.Marshal(resp)
+	if err != nil {
+		h.metrics.BadThriftFailures.Inc(1)
+		http.Error(w, "Cannot marshall Thrift to JSON", http.StatusInternalServerError)
+		return
+	}
+	if thriftEnums092 {
+		jsonBytes = h.encodeThriftEnums092(jsonBytes)
+	}
+	if err = h.writeJSON(w, jsonBytes); err != nil {
+		return
```

Comment:
not sure what should be happening in this case in HTTP, it's too late to send the status code.